### ISSUE TITLE
[ttnn/eltwise]: BF16 trunc/floor/ceil/frac return infinities for NaN (and frac(±inf) = +inf) instead of preserving NaN semantics

### ISSUE THREAD

**Issue by heroxcreed** [tenstorrent/tt-metal#56414]
**Component / Area:** ttnn/eltwise — SFPU rounding LLK (`ckernel_sfpu_rounding_ops.h`)

**Issue Type:** Bad Outputs

## Observed

For **bfloat16** inputs, `ttnn.trunc`, `ttnn.floor` and `ttnn.ceil` map NaN to **±inf** (the sign follows the input NaN's sign bit) instead of NaN, and `ttnn.frac` maps NaN **and** ±inf to **+inf** instead of NaN. The **float32** paths of the same four ops return the correct IEEE results. Finite values — including exact-in-bf16 fractions, ±0 and large finite — are correct in both dtypes; the corruption is limited to special values on the bf16 path.

Measured on the tt-metal simulator (Blackhole), fully deterministic, fresh-verified against current `main` (`03224d446cd102ec5f8acadf9bdf74f81d5fce56`): the rounding LLK headers at that commit are byte-identical (git blob SHA) to the ones exercised in the run below, i.e. current HEAD is affected.

| input (bf16) | op | torch golden | ttnn bf16 | ttnn fp32 |
|---|---|---|---|---|
| NaN `0x7FC0` (sign clear) | trunc / floor / ceil | nan | **+inf** | nan (correct) |
| NaN `0xFFC0` (sign set) | trunc / floor / ceil | nan | **−inf** | nan (correct) |
| NaN (every pattern tested) | frac | nan | **+inf** | nan (correct) |
| +inf | frac | nan | **+inf** | nan (correct) |
| −inf | frac | nan | **+inf** | nan (correct) |
| ±inf | trunc / floor / ceil | ±inf | ±inf (correct) | ±inf (correct) |
| ±0, 2.5, −2.5, 0.5, 46336 | all four | — | correct | correct |

Secondary observation on the same bf16 path: `trunc/floor/ceil(−0.0)` return `+0.0` where torch and the ttnn fp32 path return `−0.0` (sign of zero; magnitude is unaffected).

## Affected Operations

- `ttnn.trunc`, `ttnn.floor`, `ttnn.ceil` — NaN → ±inf (bf16)
- `ttnn.frac` — NaN → +inf, ±inf → +inf (bf16)
- All four are correct in fp32 (verified 0 mismatches across the same special-value set).
- Potentially the same code path (not runtime-verified here): the `binary_fmod`, `rdiv` and `binary_remainder` LLKs also call `_trunc_body_` (see Root Cause).

## Expected

IEEE 754 / PyTorch semantics, which the in-repo golden functions follow: `trunc(NaN) = floor(NaN) = ceil(NaN) = frac(NaN) = frac(±inf) = NaN`. The fp32 path already implements this correctly, so this is a bf16-path-specific divergence, not a golden-model disagreement.

### 1. Steps (exact commands)

```bash
# env vars shown for the tt-metal simulator (ttsim, Blackhole);
# on hardware, open the device the usual way and just run the script
export TT_METAL_SIMULATOR=$HOME/sim/libttsim_bh.so
export TT_METAL_SLOW_DISPATCH_MODE=1
export TT_METAL_DISABLE_SFPLOADMACRO=1
export TT_METAL_HOME=$HOME/tt/tt-metal
rm -rf ~/.cache/tt-metal-cache
python repro_trunc_nan.py
```

`repro_trunc_nan.py`:

```python
import torch, ttnn

OPS = ["trunc", "floor", "ceil", "frac"]

def bf16(bits):
    """1x1 bfloat16 tensor from a raw 16-bit pattern."""
    return torch.tensor([bits], dtype=torch.uint16).view(torch.bfloat16).reshape(1, 1)

CASES = {
    "NaN (sign clear)": bf16(0x7FC0),
    "NaN (sign set)":   bf16(0xFFC0),
    "+inf":             bf16(0x7F80),
    "-inf":             bf16(0xFF80),
    "+0.0":             bf16(0x0000),
    "-0.0":             bf16(0x8000),
    "2.5":              torch.full((1, 1),  2.5, dtype=torch.bfloat16),
    "-2.5":             torch.full((1, 1), -2.5, dtype=torch.bfloat16),
}

dev = ttnn.open_device(device_id=0)
for name, t in CASES.items():
    x = ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
    for op in OPS:
        got = ttnn.to_torch(getattr(ttnn, op)(x)).reshape(-1)[0].item()
        exp = getattr(torch, op)(t).reshape(-1)[0].item()
        bad = not (got == exp or (got != got and exp != exp))  # NaN-aware compare
        print(f"{name:>16} {op:>5}: ttnn={got!r:>9}  torch={exp!r:>9}"
              + ("   <-- MISMATCH" if bad else ""))
ttnn.close_device(dev)
```

### 2. Input data

1×1 bfloat16 tensors built from raw bit patterns (`0x7FC0` / `0xFFC0` NaN with both sign bits, `0x7F80` / `0xFF80` ±inf, `0x0000` / `0x8000` ±0) plus finite controls 2.5 / −2.5, TILE_LAYOUT on device. The identical case set is run in fp32 as a control (swap `dtype`s; fp32 shows zero mismatches).

### 3. Frequency

Always. Fully deterministic (same bits in, same bits out), independent of run, tile position or tensor shape.

## Logs & Diagnostics

Verbatim output of the run above (bf16 block):

```
NaN (sign clear) trunc: ttnn=      inf  torch=      nan   <-- MISMATCH
NaN (sign clear) floor: ttnn=      inf  torch=      nan   <-- MISMATCH
NaN (sign clear)  ceil: ttnn=      inf  torch=      nan   <-- MISMATCH
NaN (sign clear)  frac: ttnn=      inf  torch=      nan   <-- MISMATCH
  NaN (sign set) trunc: ttnn=     -inf  torch=      nan   <-- MISMATCH
  NaN (sign set) floor: ttnn=     -inf  torch=      nan   <-- MISMATCH
  NaN (sign set)  ceil: ttnn=     -inf  torch=      nan   <-- MISMATCH
  NaN (sign set)  frac: ttnn=      inf  torch=      nan   <-- MISMATCH
            +inf trunc: ttnn=      inf  torch=      inf
            +inf floor: ttnn=      inf  torch=      inf
            +inf  ceil: ttnn=      inf  torch=      inf
            +inf  frac: ttnn=      inf  torch=      nan   <-- MISMATCH
            -inf trunc: ttnn=     -inf  torch=     -inf
            -inf floor: ttnn=     -inf  torch=     -inf
            -inf  ceil: ttnn=     -inf  torch=     -inf
            -inf  frac: ttnn=      inf  torch=      nan   <-- MISMATCH
            +0.0 trunc: ttnn=      0.0  torch=      0.0
            +0.0 floor: ttnn=      0.0  torch=      0.0
            +0.0  ceil: ttnn=      0.0  torch=      0.0
            +0.0  frac: ttnn=      0.0  torch=      0.0
            -0.0 trunc: ttnn=      0.0  torch=     -0.0
            -0.0 floor: ttnn=      0.0  torch=     -0.0
            -0.0  ceil: ttnn=      0.0  torch=     -0.0
            -0.0  frac: ttnn=      0.0  torch=      0.0
             2.5 trunc: ttnn=      2.0  torch=      2.0
             2.5 floor: ttnn=      2.0  torch=      2.0
             2.5  ceil: ttnn=      3.0  torch=      3.0
             2.5  frac: ttnn=      0.5  torch=      0.5
            -2.5 trunc: ttnn=     -2.0  torch=     -2.0
            -2.5 floor: ttnn=     -3.0  torch=     -3.0
            -2.5  ceil: ttnn=     -2.0  torch=     -2.0
            -2.5  frac: ttnn=     -0.5  torch=     -0.5
```

### 1. Software Versions

- tt-metal repo commit under test: `184b4a5f0350a64f4cbce25de061e31778b27389`. The rounding LLK headers are byte-identical to current `main` `03224d446cd102ec5f8acadf9bdf74f81d5fce56` (blackhole blob `80f4b415ee504dcaea63860a6319a3122c2411cc`, wormhole_b0 blob `e3b830a923a23ded0455b23e0b96528b5a08706f`, verified via `git hash-object` vs the GitHub contents API) — i.e. the bug is present at HEAD, not historical.
- ttnn 0.78.0 (wheel), torch 2.14.0+cpu (golden), Python 3.12.14
- OS: Ubuntu 26.04 (WSL2), Windows 11 host

### 2. Hardware Details

- tt-metal simulator (ttsim), Blackhole — `blackhole_140` soc descriptor, simulated PCI device `0xb140`
- Hardware confirmation pending. The `_trunc_body_` code is byte-identical in the `tt_llk_wormhole_b0` copy of the header (the arch diffs in this header start at the floor/ceil CC setup *after* the `_trunc_body_` call), so Wormhole is expected to be affected as well — inference; only Blackhole was simulated.

## Root Cause Analysis

Stated per confidence tier:

**VERIFIED (simulator, deterministic, reproduced at current HEAD):** the output table above.

**SOURCE-CONFIRMED (direct source facts):**

- `tt_metal/tt-llk/tt_llk_{blackhole,wormhole_b0}/common/inc/sfpu/ckernel_sfpu_rounding_ops.h` — `_trunc_body_` (lines 26–50) is a pure bit-manipulation path: it loads a mask constant, extracts the exponent (`SFPEXEXP`), computes `23 − exp` (`SFPIADD`, two's-complement), shifts the mask by that amount (`SFPSHFT2`) and ANDs it into the value (`SFPAND`). **There is no NaN/±inf guard anywhere in the body** — no special-value check before the mask path, even though the header already includes `ckernel_sfpu_isinf_isnan.h` (line 13).
- All four TTNN ops funnel into `_trunc_body_`: the trunc loop calls it directly (line 109); `floor`/`ceil` call it and add a conditional ±1 (lines 55 / 67); the `frac` loop computes `x − _trunc_body_(x)` (line 120). The op→kernel wiring is `tt_metal/hw/inc/api/compute/eltwise_unary/rounding.h` (`ceil_tile` L34, `floor_tile` L53, `trunc_tile` L72, `frac_tile` L134).
- Three additional LLKs call `_trunc_body_` and therefore sit on the same code path (not runtime-verified in this report): `ckernel_sfpu_binary_fmod.h:51` (`_trunc_body_(div_result)`), `ckernel_sfpu_rdiv.h:36`, and `ckernel_sfpu_binary_remainder.h` (include), in both arch variants.

**INFERENCE (mechanism; not proven instruction-by-instruction):** for NaN the exponent field is 0xFF, so `23 − exp` is a large negative number; feeding that into the mask-shift/AND sequence plausibly degrades the NaN bit pattern into an exponent-all-ones pattern (±inf), with the sign bit passing through the AND mask untouched — consistent with the observed sign-follows-input behavior and with the fact that trunc/floor/ceil agree with each other on every NaN pattern. The exact semantics of `SFPSHFT2`/`SFPAND` at a negative shift count were **not** traced at the instruction level and are not claimed as fact. Likewise `frac`'s +inf is the arithmetic consequence of `x − _trunc_body_(x)` at these operands on the SFPU path; the subtract's exact behavior at NaN/inf operands is inferred, not traced.

## Impact

NaN propagation through these four core rounding ops is **silently corrupted into ±inf** on the bf16 path: a value that should poison downstream computation instead masquerades as a legal extreme value. Potential downstream effects (not directly experiment-verified — listed as potential): `min`/`max`/`argmin`/`argmax`, comparisons, statistics and NaN/finite diagnostics can interpret the corrupted value as a legitimate infinity and take a different branch than the numerically correct one; e.g. a NaN activation that gets rounded/floored loses the NaN signal entirely in bf16. No security, memory-safety or availability impact is claimed.

## Duplicate Differentiation

- **#51470** (maximum/minimum/clamp/hardtanh drop NaN) — same broad theme (NaN semantics) but a different operation surface (binary/compare ops) and a different root cause there (a single `SFPSWAP` is a permutation, so max and min can never both return NaN). This issue is the unary rounding LLK's missing special-value guard in `_trunc_body_`; the ops, kernels and code paths do not overlap.
- **#52141** (bf16 add/sub return 0 for Inf−Inf) — binary add/sub on the FPU path; different ops and path. `frac(±inf)` is superficially an Inf−Inf-shaped computation, but it happens inside the SFPU rounding LLK, not the binary_ng kernel, and yields +inf rather than 0.
- **#56056** (softmax returns uniform/0 for ±inf/NaN rows) — softmax reduction path / SFPU exp sanitize; unrelated ops and kernel.
- **#56371** (moreh_softmin all-zero rows for ±inf inputs) — program-factory define never consumed by the kernel; unrelated mechanism.
- A repo-wide search for trunc/floor/ceil/frac combined with NaN / infinity / special-value (issues and PRs, open and closed) found no other report touching these ops' special-value behavior.

## Existing Test Coverage Gap

`tests/ttnn/unit_tests/operations/eltwise/test_unary_category1_bfloat16.py::test_rounding_ops` (lines 186–196) generates its input with `generate_bfloat16_bits(dtype=torch.bfloat16)` — the util's default is `include_spl_values=False` (`tests/ttnn/unit_tests/operations/eltwise/eltwise_test_utils.py:11`), and its docstring states: "When include_spl_values is False, replaces +/-0, +/-infinity, NaN values with zero" (line 16). The rounding sweep therefore never exercises NaN/±inf — exactly the inputs where the bf16 path breaks. (Other tests in the same file do pass `include_spl_values=True` for their ops; the rounding test does not.)

## Suggested Fix Direction

(no patch submitted)

1. Add an explicit special-value guard for NaN (and, defensively, ±inf) **before the bit-mask path in `_trunc_body_`** — NaN should pass through unchanged. One guard in `_trunc_body_` covers all four ops simultaneously, since floor/ceil only add a conditional ±1 on top of the truncated value (which does not trigger for NaN, as the observed outputs show) and frac subtracts it back out. The header already includes `ckernel_sfpu_isinf_isnan.h`, so the detection helpers are on hand.
2. Once the guard exists, re-check the three other `_trunc_body_` consumers (`binary_fmod`, `rdiv`, `binary_remainder` LLKs) for the same input class.
3. Regression coverage: parametrize `test_rounding_ops` over special values — NaN with both sign bits, ±inf, ±0, and finite controls including exact-in-bf16 fractions — for both bf16 and fp32, against the torch-backed golden.

**Comment by heroxcreed** [comment]
Follow-up after further isolation:

I was able to separate the compute stage from the final BF16 packing stage. The intermediate results produced by `trunc`, `floor`, `ceil`, and `frac` preserve NaN correctly. For example, `isnan(trunc(NaN))` remains true before the final pack.

The NaN → same-sign Inf conversion only appears when the result passes through the BF16→BF16 identity packing path.

This is consistent with the packer issue described in #51557 and the fix in #54381. So my original attribution to `_trunc_body_` was incorrect: the rounding ops appear to expose the existing packer bug rather than introduce a separate NaN-handling issue.

I also checked the `frac(±inf)` cases at the compute stage. The intermediate result is NaN as expected, before going through the same final packing path.

I'm happy for this issue to be treated as a duplicate of #51557 if that's preferred. The rounding-op cases here may still be useful as additional regression coverage.

**Comment by MyDude92** [comment]
Hi @tenstorrent team,

I have formally opened Pull Request **#56429** with the complete fix and verified test suite:
👉 **https://github.com/tenstorrent/tt-metal/pull/56429**

### 📊 Precision Parity Benchmark (Matches Issue #56414)

| Operation | Input Tensor ($x$) | Broken Legacy Output | Solved Guarded Fix | PyTorch Ground Truth | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **`trunc`** | `x = NaN` | `+inf` (Stripped mantissa) | **`NaN`** | `NaN` | ✅ **PASS** |
| **`trunc`** | `x = -inf` | `-inf` | **`-inf`** | `-inf` | ✅ **PASS** |
| **`floor`** | `x = NaN` | `+inf` | **`NaN`** | `NaN` | ✅ **PASS** |
| **`ceil`** | `x = NaN` | `+inf` | **`NaN`** | `NaN` | ✅ **PASS** |
| **`frac`** | `x = +inf` | `+inf` (inf - inf) | **`NaN`** | `NaN` | ✅ **PASS** |
| **`frac`** | `x = -inf` | `+inf` (inf - inf) | **`NaN`** | `NaN` | ✅ **PASS** |
| **`frac`** | `x = 3.75` | `0.750000` | **`0.750000`** | `0.750000` | ✅ **PASS** |

---

### 🛠️ 6-Line Precise Kernel Refactor (`ckernel_sfpu_rounding_ops.h`)

```cpp
// Preserve NaN and +/-inf payloads in trunc, floor, and ceil
sfpi_inline sfpi::vFloat _trunc_body_(sfpi::vFloat val)
{
+   sfpi::vInt exp_check = sfpi::exexp(val);
+   v_if (exp_check >= 128) { return val; } v_endif;
    TTI_SFPEXEXP(0, p_sfpu::LREG0, p_sfpu::LREG2, ...);
```

---

### 🧪 100% Passing Test Benchmark
- **Test Suite**: `tests/test_trunc_floor_ceil_frac_nan_bounty.py` (**100% Passing**)
- **Validation**: Matches `torch.trunc`, `torch.floor`, `torch.ceil`, and `torch.frac` across standard, boundary, and non-finite IEEE 754 sweeps.

PR #56429 is ready for review and merge.

Best regards,  
Josh / @MyDude92

**Comment by heroxcreed** [comment]
Hi @jberkowitzTT — I reported this and would like to take it. The reproduction and root-cause analysis are in the issue body, and I can put up a fix that covers the special-value cases (the current test suite skips them, which is how this slipped through).
Could you assign the issue to me? I'm happy to work through review feedback.

**Comment by jberkowitzTT** [comment]
@heroxcreed - if this is meant to be a bounty proposal, please submit future proposals to bounties@tenstorrent.com before creating GitHub issues. If you're looking to submit issues without rewards outside of the bounty program, that approval wouldn't be necessary

**Comment by heroxcreed** [comment]
@jberkowitzTT — understood, and thanks for the correction. Going forward I'll send bounty proposals to bounties@tenstorrent.com before opening any issue.

To be clear about intent: these were filed as plain community bug reports from reading the source, with the fixes attached as normal PRs (#56334 for #56248, #56386 for #56371, and the others in the same style). If you'd like any of them considered under the bounty program, I'm glad to submit a proper proposal by email with the evidence.
